File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.py
import sys

cities = int(sys.stdin.readline())
is_visited = [False] * cities
costs = []
final_costs = 10000000
for i in range(cities):
    row = list(map(int, sys.stdin.readline().split()))
    costs.append(row)


def route(first: int, cur: int, depth: int, check_costs: int):
    global cities
    global final_costs

    if depth == cities and costs[cur][first] > 0:
        final_costs = min(check_costs + costs[cur][first], final_costs)
        return

    for i in range(cities):
        if not is_visited[i] and costs[cur][i] > 0:
            is_visited[i] = True
            route(first, i, depth + 1, check_costs + costs[cur][i])
            is_visited[i] = False


for i in range(cities):
    is_visited[i] = True
    route(i, i, 1, 0)
    is_visited[i] = False

sys.stdout.write(str(final_costs))

File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Map;
import java.util.StringTokenizer;

public class Main {
    private static int cities = 0;
    private static boolean[] isVisited;
    private static int[][] costs;
    private static int finalCost = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        cities = Integer.parseInt(br.readLine());
        isVisited = new boolean[cities];
        costs = new int[cities][cities];

        for (int i = 0; i < cities; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < cities; j++) {
                costs[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        for (int i = 0; i < cities; i++) {
            isVisited[i] = true;
            route(i, i, 1, 0);
            isVisited[i] = false;
        }

        System.out.println(finalCost);
    }

    private static void route(int first, int current, int depth, int checkCost) {
        if (depth == cities && costs[current][first] > 0) {
            finalCost = Math.min(checkCost + costs[current][first], finalCost);
            return;
        }

        for (int i = 0; i < cities; i++) {
            if (!isVisited[i] && costs[current][i] > 0) {
                isVisited[i] = true;
                route(first, i, depth + 1, checkCost + costs[current][i]);
                isVisited[i] = false;
            }
        }
    }
}
